src/crypto.rs

```diff
@@ -0,0 +1,158 @@
+        },
+        Algorithm::RS256 | Algorithm::RS384 | Algorithm::RS512 => {
+            let ring_alg = match algorithm {
+                Algorithm::RS256 => &signature::RSA_PKCS1_SHA256,
+                Algorithm::RS384 => &signature::RSA_PKCS1_SHA384,
+                Algorithm::RS512 => &signature::RSA_PKCS1_SHA512,
+                _ => unreachable!(),
+            };
+            // Taken from Ring doc https://briansmith.org/rustdoc/ring/signature/index.html
+            let key_pair = Arc::new(
+                signature::RSAKeyPair::from_der(
+                    untrusted::Input::from(key)
+                ).map_err(|_| ErrorKind::InvalidKey)?
+            );
+            let mut signing_state = signature::RSASigningState::new(key_pair)
+                .map_err(|_| ErrorKind::InvalidKey)?;
+            let mut signature = vec![0; signing_state.key_pair().public_modulus_len()];
+            let rng = rand::SystemRandom::new();
+            signing_state.sign(ring_alg, &rng, data.as_bytes(), &mut signature)
+                .map_err(|_| ErrorKind::InvalidKey)?;
+
+            Ok(base64::encode_config(
+                signature.as_ref(),
+                base64::URL_SAFE_NO_PAD
+            ))
+        },
+    }
+}
+
+/// Encode the claims passed and sign the payload using the algorithm from the header and the key
+pub fn encode<T: Serialize>(header: Header, claims: &T, key: &[u8]) -> Result<String> {
+    let encoded_header = to_jwt_part(&header)?;
+    let encoded_claims = to_jwt_part(&claims)?;
+    let payload = [encoded_header.as_ref(), encoded_claims.as_ref()].join(".");
+    let signature = sign(&*payload, key.as_ref(), header.alg)?;
+
+    Ok([payload, signature].join("."))
+}
+
+/// Compares the signature given with a re-computed signature for HMAC or using the public key (`key`)
+/// for RSA
+pub fn verify(signature: &str, payload: &str, key: &[u8], algorithm: Algorithm) -> Result<bool> {
+    match algorithm {
+        Algorithm::HS256 | Algorithm::HS384 | Algorithm::HS512 => {
+            // we just re-sign the data with the key and compare if they are equal
+            let signed = sign(payload, key, algorithm)?;
+            Ok(verify_slices_are_equal(signature.as_ref(), signed.as_ref()).is_ok())
+        },
+        Algorithm::RS256 | Algorithm::RS384 | Algorithm::RS512 => {
+            // we use ring to verify using the public key given
+            let verification_alg = match algorithm {
+                Algorithm::RS256 => &signature::RSA_PKCS1_2048_8192_SHA256,
+                Algorithm::RS384 => &signature::RSA_PKCS1_2048_8192_SHA384,
+                Algorithm::RS512 => &signature::RSA_PKCS1_2048_8192_SHA512,
+                _ => unreachable!(),
+            };
+            let res = signature::verify(
+                verification_alg,
+                untrusted::Input::from(key),
+                untrusted::Input::from(payload.as_bytes()),
+                untrusted::Input::from(signature.as_bytes()),
```

Comment:
Or in ring test format: https://github.com/Keats/ring/commit/fb38c792b5a3a53acda94a59a1d5b98301bf3b49 fails:

```
---- rsa::verification::tests::test_signature_rsa_pkcs1_verify stdout ----
	thread 'rsa::verification::tests::test_signature_rsa_pkcs1_verify' panicked at 'assertion failed: public_key.read_all(error::Unspecified,
                    |input|
                        {
                            der::nested(input, der::Tag::Sequence,
                                        error::Unspecified,
                                        |input|
                                            {
                                                let _ =
                                                    try!(der ::
                                                         positive_integer (
                                                         input ));
                                                let _ =
                                                    try!(der ::
                                                         positive_integer (
                                                         input ));
                                                Ok(())
                                            })
                        }).is_ok()', src/rsa/verification.rs:192
note: Run with `RUST_BACKTRACE=1` for a backtrace.
src/rsa/rsa_pkcs1_verify_tests.txt: Test panicked.
Digest = SHA256
Key = 30820122300d06092a864886f70d01010105000382010f003082010a0282010100c9113aac7b8d47441b1cedc7dcab76a4e28656142a1995c89ce76e40dc57cee2a5bd04cb513bf8978b20821e7f098622fdcbc8f925d54fd90f592297c495c15ddff82e4bdc3ee51a901a0091f87e7a2155321d95ad4c96ca3dcc165d074d517d2b04572c07309111224b79e94e11d1c88c6ecb464c7997f154be5aacc870d524442c1f07a067c6fc0b47f3d04813d8c304767d74b7a52bd6b5f38cc07fc2f0a0f2f1bc96f7225e679dca8f7127ca0c3a1d30504831ce254330ca2f982f9a25cb5c1d4018b9bc2818df13cb372f9c6a8bec94a1dfa3f0cb6f223f35d9d912e1032245537f6f2da1dd963c2d8546aea6576537209f6ba39fcb8a8d72d9543e53750203010001
Msg (unconsumed) = 65794a30655841694f694a4b563151694c434a68624763694f694a53557a49314e694973496d707264534936626e56736243776961326c6b496a7075645778734c434a344e5855694f6d353162477773496e673164434936626e56736248302e65794a7a645749694f694a69514749755932397449697769593239746347467565534936496b4644545555696651
Sig (unconsumed) = 6b53f4f56037d0b4d7c995792ad39cd88fc073c874a38ede547742ff54b725f6ab52ce7babbb05a0bfb8d3b6a214dac247e41353cd5b727c98889f8427c6595ef756be982ae8353bfdf1049ccd7d8783e1834660294de282fe579e413d257327d179390ea51443536fe16c8c8c933bce455b6249018a4df831d0addfd95d72527d2915cbc300acbda2535bc3c6ceff5ef25a1f13b625f9dfb5d6a53da536110fb4f4a3cdeb20fb04a3f2e6818db781a9566d99a5a83a445dd4284cd4fe83079ef55e0ce12f632c937cacf2601e54858c3b1a1dc39d243468c0d3ae3c7d4ecf6c1b7777e11371fd4a3f3690d441af17fe8644d0a08b19f6f10b7c3eb7f79718be
Result (unconsumed) = P
thread 'rsa::verification::tests::test_signature_rsa_pkcs1_verify' panicked at 'Test failed.', src/test.rs:284
```

The keys have been generated with the following commands (and are in that branch in the `tests` directory too):

```md
# Generating RSA keys

Using `openssl`

## PEM
`openssl genrsa -out private_rsa_key.pem 2048`

Getting public key:
`openssl rsa -in private_rsa_key.pem -outform PEM -pubout -out public_rsa_key.pem`

## DER
Same as PEM but replace `PEM` by `DER`.
`openssl rsa -in private_rsa_key.pem -outform DER -pubout -out public_rsa_key.der`

## Converting private PEM to DER
`openssl rsa -in private_rsa_key.pem -outform DER -out private_rsa_key.der`

## Converting private DER to PEM
`openssl rsa -in private_rsa_key.der -inform DER -outform PEM -out private_rsa_key.pem`

```